docs/devGuide/migratingToTypeScript.md

```diff
@@ -0,0 +1,131 @@
+
+   - If you are trying to figure out the types for parameters, you can infer from how they are used in the function, or see how the function is called in other files.
+
+   - Avoid using `any` as best as you can.
+
+4. **Make sure everything is in order** by running `npm run test` in the root directory.
+
+   The `core` package test directly tests `.ts` files (powered by `ts-jest`), and the `cli` package test uses the compiled files from `core`. Therefore, passing the `cli` tests means that we are ensured the compiled files work as expected.
+
+5. **Stage the adapted files and the typings from Step 1** for commit. Verify the changes before committing.
+
+6. **Commit the changes** with the message `Adapt <file-or-folder> to TypeScript`.
+   <br>
+   For example, `Adapt core/src/html to TypeScript`.
+
+You are now ready to create a pull request for the changes to the repository.
+
+## Examples of Migrated Works
+
+You can see these pull requests to observe the finished migration works:
+
+- [Pull Request #1877: Adopt TypeScript for core package](https://github.com/MarkBind/markbind/pull/1877)
+
+## Explanation Notes
+
+### Why do we need separate "Rename" and "Adapt" commits
+
+We want to keep the file history intact throughout this migration, meaning that we should still be able to see the change history of the `.ts` files even from the time when they were still in `.js`.
+
+When we rename files, we have to keep the Git similarity index between the old file and the new file to be above a certain threshold (50%) to keep the history, or else it will be regarded as a different file with its own history.
+
+Combining the two steps into one - that is, adapting files immediately after renaming - will lower the similarity index, therefore making it possible for the files, especially small ones or those that need a lot of changes when adapting, to hit below the similarity threshold and lose their history.
+
+The solution to this is to make the rename and adapt commits separate, in order for Git to recognize first that the file has changed into `.ts`, and so the changes are compared not against the old `.js` file, but to the renamed `.ts` file instead.
+
+### Why do we need a merge commit
+
+Even if the migration developer has kept the history intact through the separate "Rename" and "Adapt" commits, this is only intact at their working branch. At the end of the day, the totality of the changes in the working branch is compared against the target branch, in which the same consideration would apply.
+
+If we do the usual squash commit, the two separate commits are combined into one, and so the problem with losing history occurs again, but now in the context of the target branch. Again, the solution is to keep those two commits separate, even in the target branch, and therefore what we must merge the PR with a merge commit to achieve this.
```

Comment:
👀 
Dosen't a merge PR still squash everything into one commit in the target branch? (hence the need for `rename-threshold`)

But preserving the history in the working branch, from which git can _possibly_ infer the rename.
(In contrast, squash removes the information from the working branch completely)

Reply:
Hmm yeah, now that I think about it, I felt like I was only basing it like a fast-forward merge rather than an actual merge commit haha. Will update into it.